### PR TITLE
CSHARP-4627: Support Union via $unionWith in LINQ3

### PR DESCRIPTION
**Pull request by sanych-sun**

### CODE REVIEW

**Review by rstam — CHANGES_REQUESTED** [review]
Good work. I have some non-trivial refactoring suggestions based on my experience writing LINQ3 translators. Let's meet on Zoom tomorrow to discuss.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToPipelineTranslators/ExpressionToPipelineTranslator.cs [hunk 1]
```diff
@@ -67,6 +67,8 @@ public static AstPipeline Translate(TranslationContext context, Expression expre
                     return TakeMethodToPipelineTranslator.Translate(context, methodCallExpression);
                 case "Where":
                     return WhereMethodToPipelineTranslator.Translate(context, methodCallExpression);
+                case "Union":
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToPipelineTranslators/ExpressionToPipelineTranslator.cs:68]: Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Translators/ExpressionToPipelineTranslators/UnionMethodToPipelineTranslatorTests.cs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+        [Fact]
+        public void Union_should_combine_collections()
+        {
+            var queryable = _firstCollection
+                .AsQueryable()
+                .Union(_secondCollection.AsQueryable());
+
+            var stages = Translate(_firstCollection, queryable);
+            AssertStages(stages, "{ $unionWith : { coll : 'partners', pipeline : [] } }");
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Translators/ExpressionToPipelineTranslators/UnionMethodToPipelineTranslatorTests.cs]: Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Translators/ExpressionToPipelineTranslators/UnionMethodToPipelineTranslatorTests.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+            var items = queryable.ToList();
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Translators/ExpressionToPipelineTranslators/UnionMethodToPipelineTranslatorTests.cs]: Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Translators/ExpressionToPipelineTranslators/UnionMethodToPipelineTranslatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+        [Fact]
+        public void Union_should_combine_collection_with_itself()
+        {
+            var queryable = _firstCollection
+                .AsQueryable()
+                .Union(_firstCollection.AsQueryable());
+
+            var stages = Translate(_firstCollection, queryable);
+            AssertStages(stages, "{ $unionWith : { coll : 'clients', pipeline : [] } }");
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Translators/ExpressionToPipelineTranslators/UnionMethodToPipelineTranslatorTests.cs]: Done

File: src/MongoDB.Driver/Linq/Linq3Implementation/ExtensionMethods/ExpressionExtensions.cs [hunk 5]
```diff
@@ -44,5 +45,19 @@ public static TValue GetConstantValue<TValue>(this Expression expression, Expres
             var message = $"Expression must be a constant: {expression} in {containingExpression}.";
             throw new ExpressionNotSupportedException(message);
         }
+
+        public static object Evaluate(this Expression expression)
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/ExtensionMethods/ExpressionExtensions.cs:49]: As a general rule we keep methods in alphabetical order.

**Review by rstam — CHANGES_REQUESTED** [review]
99% there

File: src/MongoDB.Driver/Linq/Linq3Implementation/ExtensionMethods/ExpressionExtensions.cs (hunk 5 above)
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/ExtensionMethods/ExpressionExtensions.cs]: Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Translators/ExpressionToPipelineTranslators/UnionMethodToPipelineTranslatorTests.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+                new Company { Id = 1, Name = "first client" },
+                new Company { Id = 2, Name = "second client" },
+                new Company { Id = 3, Name = "third client" });
+
+            _secondCollection = CreateCollection("partners",
+                new Company { Id = 4, Name = "partner" },
+                new Company { Id = 5, Name = "another partner" });
+        }
+
+
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Translators/ExpressionToPipelineTranslators/UnionMethodToPipelineTranslatorTests.cs]: Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Translators/ExpressionToPipelineTranslators/UnionMethodToPipelineTranslatorTests.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        [Fact]
+        public void Union_should_combine_filtered_collections()
+        {
+            var queryable = _firstCollection
+                .AsQueryable()
+                .Where(c => c.Name.StartsWith("second"))
+                .Union(_secondCollection.AsQueryable()
+                    .Where(c => c.Name.StartsWith("another")));
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Translators/ExpressionToPipelineTranslators/UnionMethodToPipelineTranslatorTests.cs]: Done

**Review by rstam — APPROVED** [review]
LGTM

Looks great!

**Pull request merged**